### PR TITLE
[Python] add to_byte_buffer to literal classes

### PR DESCRIPTION
**Pull request by jun-he**

### CODE REVIEW

**Comment by jun-he** [comment]
FYI @rymurr @TGooch44, here is the change related to https://github.com/apache/iceberg/pull/414

File: python/iceberg/api/expressions/literals.py
Comment (rymurr) [inline python/iceberg/api/expressions/literals.py:263]: I wonder if it would be cleaner to add the `type_id` as a `__init__` param in `BaseLiteral`?

**Review by rymurr — COMMENTED** [review]
Looks great @jun-he 

My only comment would be if its cleaner to put the type_id in the constructor and not have every `Literal` implement the `type_id` method. I don't feel strongly either way, up to you.

File: python/iceberg/api/types/conversions.py
Comment (TGooch44) [inline python/iceberg/api/types/conversions.py:89]: Since you're touching this code anyway, it may be more appropriate to raise either TypeError or NotImplementedError here.  There are a number of exceptions in this codebase that raise Runtime but should raise something more specific.

File: python/iceberg/api/expressions/literals.py
Comment (TGooch44) [reply python/iceberg/api/expressions/literals.py:263]: That makes the most sense to me.  Since it will cause a number of changes in unrelated files, what do you think about opening an issue on it, and handling in a separate PR?

**Comment by TGooch44** [comment]
Overall, looks good to me.

File: python/iceberg/api/expressions/literals.py
Comment (rymurr) [reply python/iceberg/api/expressions/literals.py:263]: Why would it affect other files? To me it would just make this diff smaller: add `type_id` parameter to `__init__` and return it instead of the `NotImplementedError` in `BaseLiteral.type_id`, then we don't have to implement `def type_id(self):` in every subclass.

But yeah, not overly precious about it. Can def be done in a followup.

Comment (TGooch44) [reply python/iceberg/api/expressions/literals.py:263]: I think I misunderstood what you were proposing.  Let me walk that back, haha.  You're just saying in the child-class pass into the super constructor the type_id of the child and then just implement a single type_id attr in the BaseLiteral.  I agree that does seem better to me and is appropriate to handle within this PR.  

I'm also going to go ahead and finish my coffee before reading any more PRs.

Comment (rymurr) [reply python/iceberg/api/expressions/literals.py:263]: > I'm also going to go ahead and finish my coffee before reading any more PRs.

:rofl:

Comment (jun-he) [reply python/iceberg/api/expressions/literals.py:263]: 👍  thanks for the comments. SGTM

File: python/iceberg/api/types/conversions.py
Comment (jun-he) [reply python/iceberg/api/types/conversions.py:89]: Yep, update accordingly.

**Comment by jun-he** [comment]
@TGooch44 @rymurr Thanks for the comments and updated the PR accordingly. 

Additionally, looks like the `to_byte_buff_mapping` and `from_byte_buff_mapping` are not symmetric and I am going to send another PR to keep them consistent.

**Review by rymurr — APPROVED** [review]
Looks great @jun-he thanks a lot!

I will wait for @TGooch44 to wake up and have a coffee before merging ;-)

**Review by TGooch44 — APPROVED** [review]
@rymurr I'm still not seeing merge access, so in the interest of expediency, can you merge this while I get my write access sorted out?

**Pull request merged**
